### DISCUSSION TITLE
How do I match multiple lines of code?

### DISCUSSION THREAD

**Discussion by raffomania** [ast-grep/ast-grep#690] [Q&A] [answered]
I would like to match two lines from a python codebase using FastAPI:

```py
router = APIRouter(prefix="/user", dependencies=[Depends(get_current_user)])
tags = ["user"]
```

I've tried to use this command:
`ast-grep run --lang python --pattern 'router = APIRouter($$$ARGS) tags = $$$TAGS'`

Resulting in the following error:
```
Error: Cannot parse query as a valid pattern.
Help: The pattern either fails to parse or contains error. Please refer to pattern syntax guide.
See also: https://ast-grep.github.io/guide/pattern-syntax.html

✖ Caused by
╰▻ Multiple AST nodes are detected. Please check the pattern source `router = APIRouter($$$ARGS) tags = $$$TAGS`.
```

In the playground, using the following pattern:
```
router = APIRouter($$$ARGS)
tags = $$$TAGS
```

can be parsed, but it doesn't match the code above. What is the correct syntax for matching newlines and/or whitespace?

**Comment by HerringtonDarkholme** [comment, marked as answer]
Hi, thanks for giving ast-grep a try.

It is not possible to match two nodes at one time now.

Alternatively, you can try relational rule https://ast-grep.github.io/guide/rule-config/relational-rule.html

**Reply by raffomania** [reply]
Thanks for clarifying!

**Comment by gmile** [comment]
@HerringtonDarkholme has anything changed in this regard since? I'd love to be able to refactor 2 lines of code into 1.

For example in my Elixir tests, I want to find all code like this:

```elixir
conn = delete(conn, contract_path(conn, :delete, contract.id))

assert json_response(conn, 401)
```

and change it to code like this:

```elixir
conn
|> delete(contract_path(conn, :delete, contract.id))
|> json_response(401)
```

Using relational rule works:

```elixir
id: test-conn-to-pipe-syntax
language: elixir
rule:
  pattern: conn = $METHOD(conn, $PATH)
  precedes:
    pattern: assert json_response(conn, $STATUS)
fix: |
  conn
  |> $METHOD($PATH)
  |> json_response($STATUS)
```

However, the update won't remove the `assert json_response(conn, 401)`, so the diff would look like this:

```diff
@@ -1282,7 +1289,9 @@ defmodule MyApp.ContractControllerTest do
     test "returns 401 for unauthenticated request", %{conn: conn} do
       contract = insert(:contract)

-      conn = delete(conn, contract_path(conn, :delete, contract.id))
+      conn
+      |> delete(contract_path(conn, :delete, contract.id))
+      |> json_response(401)

       assert json_response(conn, 401)
     end
```

**Reply by gmile** [reply]
Might be related: https://github.com/ast-grep/ast-grep/issues/1985

**Reply by HerringtonDarkholme** [reply]
for this case you can use https://ast-grep.github.io/guide/rewrite-code.html#expand-the-matching-range `expandEnd` to include more AST nodes

**Reply by gmile** [reply]
@HerringtonDarkholme thank you! This helped me unblock the replacement I was intending to do. I've written down about my short journey here: https://ievgenpyrogov.com/cleaning-up-code-using-ast-grep/

**Reply by HerringtonDarkholme** [reply]
Thanks for the nice writeup!!
